File: src/scene/Scene.java
/* (C)2023 */
package scene;

import app.Salient;
import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.graphics.Cubemap;
import com.badlogic.gdx.graphics.PerspectiveCamera;
import com.badlogic.gdx.graphics.Texture;
import io.DirectoryMappings;
import io.Serializer;
import net.mgsx.gltf.scene3d.attributes.PBRCubemapAttribute;
import net.mgsx.gltf.scene3d.attributes.PBRTextureAttribute;
import net.mgsx.gltf.scene3d.lights.DirectionalLightEx;
import net.mgsx.gltf.scene3d.scene.SceneManager;
import net.mgsx.gltf.scene3d.scene.SceneSkybox;
import net.mgsx.gltf.scene3d.utils.IBLBuilder;
import org.yunghegel.gdx.gizmo.core.utility.CompassGizmo;
import org.yunghegel.gdx.scenegraph.scene3d.BaseScene;
import org.yunghegel.gdx.ui.widgets.viewport.ViewportWidget;
import org.yunghegel.gdx.utils.graphics.Grid;
import project.Project;
import scene.context.env.EditorEnvironment;
import sys.profiling.Profile;
import util.SceneUtils;

public class Scene extends BaseScene implements ViewportWidget.Renderer {

    public String name;
    public String path;
    public String uid = "-1";

    private SceneGraph sceneGraph;
    public Project project;

    public EditorEnvironment env = new EditorEnvironment();
    public PerspectiveCamera camera = Salient.INSTANCE.getCamera();
    public Grid grid;
    public SceneSkybox skybox;
    public SceneManager sceneManager;
    public static CompassGizmo compassGizmo;

    public Scene(String name, Project project) {
        sceneGraph = new SceneGraph(this);
        this.name = name;
        this.project = project;
        path = DirectoryMappings.INSTANCE.getScenePath(project.getName(), name);
        float[] subdivisions = new float[] {0.25f, 0.125f};
        grid = new Grid(0.2f, 300, 0.5f, subdivisions);
        Cubemap customSkyboxCubemap = SceneUtils.createCubemapDirectionFormat("editor/", "editor/");
        skybox = new SceneSkybox(customSkyboxCubemap);
        sceneManager = new SceneManager();
        sceneManager.setCamera(Salient.INSTANCE.getCameraController().getPerspectiveCamera());

        IBLBuilder iblBuilder = IBLBuilder.createOutdoor(new DirectionalLightEx());
        Texture brdfLUT = new Texture(Gdx.files.classpath("net/mgsx/gltf/shaders/brdfLUT.png"));
        Cubemap environmentCubemap = iblBuilder.buildEnvMap(1024);
        Cubemap diffuseCubemap = iblBuilder.buildIrradianceMap(256);
        Cubemap specularCubemap = iblBuilder.buildRadianceMap(10);

        sceneManager.setAmbientLight(.1f);
        sceneManager.environment.set(
                new PBRTextureAttribute(PBRTextureAttribute.BRDFLUTTexture, brdfLUT));
        sceneManager.environment.set(PBRCubemapAttribute.createSpecularEnv(specularCubemap));
        sceneManager.environment.set(PBRCubemapAttribute.createDiffuseEnv(diffuseCubemap));

        sceneManager.setSkyBox(skybox);

        Gdx.gl.glEnable(Gdx.gl.GL_DEPTH_TEST);
        Gdx.gl.glDepthFunc(Gdx.gl.GL_LEQUAL);
        Gdx.gl.glDepthMask(true);

        Gdx.gl.glEnable(Gdx.gl.GL_CULL_FACE);
        Gdx.gl.glCullFace(Gdx.gl.GL_BACK);

        Gdx.gl.glEnable(Gdx.gl.GL_BLEND);
        Gdx.gl.glBlendFunc(Gdx.gl.GL_SRC_ALPHA, Gdx.gl.GL_ONE_MINUS_SRC_ALPHA);

        compassGizmo =
                new CompassGizmo(
                        Salient.INSTANCE.getInputManager().getInputMultiplexer(),
                        Salient.INSTANCE.getModelBatch(),
                        Salient.INSTANCE.getCamera(),
                        Salient.INSTANCE.getViewport(),
                        0);
    }

    public Scene(Project project) {
        this("New Scene", project);
    }

    @Override
    public void update(float delta) {
        sceneGraph.update(delta);
        sceneManager.update(delta);
        Salient.INSTANCE.getCameraController().getPerspectiveCameraController().update(delta);
        camera.update();
        compassGizmo.update();
        //        Gdx.input.setInputProcessor(Salient.INSTANCE.getCameraController());

    }

    @Override
    public void render(float delta) {

        sceneManager.render();
        grid.render(camera);
        sceneGraph.render(delta);
        Salient.INSTANCE.getCameraController().getPerspectiveCameraController().render();
    }

    @Override
    public void resize(int width, int height) {}

    @Override
    public void dispose() {}

    public static Scene createDefault(Project proj) {
        return new Scene("New Scene", proj);
    }

    @Profile
    public void save() {
        Serializer.INSTANCE.serializeScene(this);
    }
}

File: src/scene/Scene.kt
package scene

import app.Salient
import com.badlogic.gdx.Gdx
import com.badlogic.gdx.graphics.GL20
import com.badlogic.gdx.graphics.OrthographicCamera
import com.badlogic.gdx.graphics.PerspectiveCamera
import com.badlogic.gdx.graphics.Pixmap
import com.badlogic.gdx.graphics.g3d.ModelBatch
import com.badlogic.gdx.utils.viewport.ScreenViewport
import com.crashinvaders.vfx.VfxManager
import com.crashinvaders.vfx.effects.FxaaEffect
import events.scene.GameObjectDeselectedEvent
import events.scene.GameObjectSelectedEvent
import input.EditorCamera
import input.InputManager
import io.Serializer
import ktx.collections.GdxArray
import org.yunghegel.gdx.scenegraph.scene3d.BaseScene
import org.yunghegel.gdx.ui.widgets.viewport.ViewportWidget
import project.Project
import scene.graph.GameObject
import scene.graph.SceneGraph
import scene.systems.SceneIterator
import tools.ToolManager

/**
 * Default subclass of the [BaseScene] class - this is the main scene class used by the editor.
 *
 * Render context (environment) and configuration is provided by the [SceneContext] class.
 * Handling of the actual rendering is done by the [SceneRenderer] class which is configured by the [SceneContext].
 *
 * Only required parameter is [project], a scene must be associated with a project.
 */
class Scene( var name: String = "New Scene", val project: Project, var uid : String = "-1") : BaseScene(name,project.sceneFolder.path()+'/'+name.plus(".scene")), ViewportWidget.Renderer,GameObjectSelectedEvent.Listener,GameObjectDeselectedEvent.Listener {




    var perspectiveCamera : PerspectiveCamera = Salient.inject()
    var orthographicCamera : OrthographicCamera = Salient.inject()
    var controller : EditorCamera = Salient.inject()
    var inputs: InputManager = Salient.inject()
    var viewport : ScreenViewport = Salient.inject()

    var modelBatch : ModelBatch
    var depthBatch : ModelBatch

    var sceneRenderer : SceneRenderer
    var sceneGraph : SceneGraph
    var sceneContext : SceneContext

    var sceneIterators: GdxArray<SceneIterator> = GdxArray()

    var selectedGameObjects: GdxArray<GameObject> = GdxArray()
    var selectedGameObject: GameObject? = null

    lateinit var effectsManager : VfxManager


    lateinit var data : SceneData

    init {
        sceneRenderer = SceneRenderer()
        sceneGraph = SceneGraph(this)
        sceneContext = SceneContext()

        sceneRenderer.configureFromScene(this)

        modelBatch = sceneRenderer.batch
        depthBatch = sceneRenderer.depthBatch


        effectsManager= VfxManager(Pixmap.Format.RGBA8888)
        effectsManager.addEffect(FxaaEffect())

        Salient.registerListener(this)
    }

    fun hasSelection(): Boolean {
        return selectedGameObjects.size > 0
    }


    override fun dispose() {
        sceneRenderer.dispose()
        sceneContext.dispose()
    }

    override fun update(delta: Float) {
        sceneGraph.update(delta)
        sceneRenderer.update(delta)
        Salient.compass.update()

        perspectiveCamera.update()
    }

    override fun render(delta: Float) {
        Gdx.gl.glClear(GL20.GL_COLOR_BUFFER_BIT or GL20.GL_DEPTH_BUFFER_BIT)
        sceneRenderer.render()
        sceneRenderer.renderSceneContext(delta,sceneContext)
        sceneRenderer.renderSceneGraph(delta,sceneGraph)
        controller.perspectiveCameraController.render()
        ToolManager.renderTool()
    }

    override fun resize(width: Int, height: Int) {

    }

   companion object {
       fun createDefault(project: Project): Scene {
           return Scene("Default Scene", project)
       }

   }

    fun save() {
        Serializer.serializeScene(this)
    }

    fun getRenderer(): ViewportWidget.Renderer {
        return ViewportRenderer(this)
    }

    class ViewportRenderer(val scene: Scene) : ViewportWidget.Renderer {
        override fun render(delta: Float) {
            scene.render(delta)
        }
    }

    override fun onGameObjectDeselected(event: GameObjectDeselectedEvent) {
        selectedGameObjects.removeValue(event.gameObject,true)
    }

    override fun onGameObjectSelected(event: GameObjectSelectedEvent) {
        if(!event.inclusive) {
            selectedGameObjects.clear()
        }
        selectedGameObjects.add(event.gameObject)
        selectedGameObject = event.gameObject
    }

}